### PR TITLE
Feature/jb 383 tide metrics

### PR DESCRIPTION
**Pull request by sei-wshaw**
Still in work in progress but most of it's in there.

### CODE REVIEW

File: juneberry/metrics/metrics_manager.py [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+                formatter_fqcn = i["formatter"]["fqcn"]
+                formatter_kwargs = i["formatter"].get("kwargs", None)
+                entry.formatter = loader.construct_instance(formatter_fqcn, formatter_kwargs)
+
+            self.config.append(entry)
+
+    def __call__(self, anno: Dict, det: Dict) -> Dict[str, Any]:
+        """
+        Compute metrics given annotations and detections in dicts.
+        :param anno_file: Annotations dict
```
Comment (sei-amellinger) [inline juneberry/metrics/metrics_manager.py:91]: Params don't match signature.

We might want to mention these a coco file format.

File: juneberry/metrics/metrics_manager.py [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        """
+        with tempfile.NamedTemporaryFile(mode="w+") as anno_file, tempfile.NamedTemporaryFile(mode="w+") as det_file:
+            # Write the anno_data and det_data dicts to temporary files; the Brambox
+            # load methods take files as input.
+            json.dump(anno_data, anno_file)
+            json.dump(det_data, det_file)
+            anno_file.flush()
+            det_file.flush()
+            # Load anno_file and det_file via brambox into DataFrames
+            anno_df = bb.io.load("anno_coco", anno_file.name, parse_image_names=False)
```
Comment (sei-amellinger) [inline juneberry/metrics/metrics_manager.py:151]: Have we looked at using the coco parser directly?

https://eavise.gitlab.io/brambox/_modules/brambox/io/parser/annotation/_coco.html#CocoParser

File: juneberry/metrics/metrics_manager.py [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            # NOTE: Loading the detections requires access to the annotations file. That's why we load
+            # detections and annotations in the same method; so that the TemporaryFiles anno_file
+            # and det_file both exist for this call.
+            det_df = bb.io.load("det_coco",
+                                det_file.name,
+                                class_label_map=MetricsManager._get_class_label_map(anno_file.name))
+            return anno_df, det_df
+
+    @staticmethod
+    def _get_class_label_map(anno_file: str) -> List[str]:
```
Comment (sei-amellinger) [inline juneberry/metrics/metrics_manager.py:161]: Can you promote this to common function in coco_utils?

Comment (sei-wshaw) [reply juneberry/metrics/metrics_manager.py:161]: Sure, np.

File: juneberry/metrics/metrics_manager.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+        Create a metrics plugin instance for each entry in the config.
+        :param config: list of metrics plugins from the config file
+        :param opt_args: optional arguments passed to this metrics manager
+        """
+        self.config = []
+
+        # For each metrics plugin entry in the config, instantiate a metrics plugin object
+        # (and associated formatter, if given) and add to the list of metrics
+        for i in config:
+            if "fqcn" not in i:
```
[the remark this answers is not in the document]
Comment (sei-wshaw) [reply juneberry/metrics/metrics_manager.py]: sounds good, converted to List[Metrics]

**Review by sei-amellinger — APPROVED** [review]

**Pull request merged**